### ISSUE TITLE
Proposal: Add a DigestBuilder

### ISSUE THREAD

**Issue by mehcode** [RustCrypto/hashes#61]
Instead of this:

```rust
let foo = {
    let mut sh = Sha1::default();
    sh.input(&bar);
    sh.input(&baz);
    sh.input(&buz);
    sh.result()
};
```

It would allow for this:

```rust
let foo = Sha1::builder()
    .input(&bar)
    .input(&baz)
    .input(&buz)
    .result()
```

This could be added entirely in the `digest` crate.

**Comment by burdges** [comment]
It's likely okay doing this with `#[inline(always)]` methods and using them like you did, but if users pass hashes around by value because they like this API then eventually the rustc starts copying them, which hurts performance and leaves secret information more places in memory.  

In principle, `input` could pass through `&mut self` so that this worked
```
let mut sh = Sha1::default();
let foo = sh.input(&bar)
    .input(&baz)
    .input(&buz)
    .result();
```

**Comment by mehcode** [comment]
Yeah. Just changing `input` to return `&mut Self` makes this pattern work just fine to me. I'm concerned of the trait breakage but it looks like `digest 0.8` already has some breakage there for anyone to deal with.

**Comment by tarcieri** [comment]
> Yeah. Just changing input to return &mut Self makes this pattern work just fine to me

As a `Digest` power user, I've pondered this same change to the API, and would be a fan. I think most usages wouldn't be affected by this change, it has some neat advantages (especially in conjunction with iterator APIs), and avoiding the extra type with a simple change to the trait API would be a plus in my mind.

**Comment by newpavlov** [comment]
Good idea to add support for chained calls!

On the first glance having a separate `DigestBuilder<D: Digest>` struct looks like a better approach to me for several reasons:
- You don't need to create temporary variable `sh` and can write everything in one chain.
- `Digest::input` stays absolutely the same, so less chances to break something on `digest v0.8` migration.
- I would like to keep `result` a self-consuming method to prevent potential misuses (note that in v0.8 there is a separate `result_reset` method), keeping this in mind @burdges snippet will not work here.
- Having a separate builder type will be probably less surprising for users.

The code can look like this:
```rust
struct DigestBuilder<D> {
    hasher: D,
}

impl<D: Digest> DigestBuilder<D> {
    // there will be also `Digest::builder() -> DigestBuilder<D>` method
    fn new(hasher: D) -> Self {
        DigestBuilder { hasher }
    }

    fn input(self, data: &[u8]) -> Self {
        self.hasher.input(data);
        self
    }
    
    fn result(self) -> GenericArray<u8, D::OutputSize> {
        self.hasher.result()
    }
    
    fn build(self) -> D {
        self.hasher
    }
}
```

What to do you think?

**Comment by mehcode** [comment]
I already tested a patch that only changes it to return `&mut Self` and my original example would work with a change from `::builder()` to `::default()`.

However, a separate struct gives us another benefit (beyond what @newpavlov is describing). We could enhance the usability of `.input` without breaking the `Digest` trait too much.

```rust
fn input<D: AsRef<[u8]>>(self, data: D) -> Self { ... }
```

**Comment by newpavlov** [comment]
I think using `AsRef<[u8]>` instead of `&[u8]` shouldn't cause any problems, so we could update `Digest` in v0.8 directly.

**Comment by newpavlov** [comment]
Instead of a separate `DigestBuilder` struct I've added a separate method [`chain`](https://github.com/RustCrypto/traits/blob/digest_v0.8/digest/src/lib.rs#L37) (not the best name I guess, but I don't have better ideas right now) which is proxied in the `Digest` trait. So in v0.8 you will be able to write like this:
```Rust
let result = Sha1::new()
    .chain(b"Hello")
    .chain(" ")
    .chain("world")
    .result();
```

And it will work with `VariableOutput` and `FixedOutput` implementators as well.

**Comment by newpavlov** [comment]
Fixed with `digest v0.8`.
